crates/ruff_linter/src/rules/pylint/rules/repeated_keywords.rs

```diff
@@ -0,0 +1,93 @@
+use std::collections::HashSet;
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{Expr, ExprDict, ExprStringLiteral, Keyword};
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for repeated keyword arguments passed to a function call
+///
+/// ## Why is this bad?
+/// Python does not allow for multiple values to be assigned to the same
+/// keyword argument in a single function call.
+///
+/// ## Example
+/// ```python
+/// func(1, 2, c=3, **{"c": 4})
+/// ```
+///
+/// Use instead:
+/// ```python
+/// func(1, 2, **{"c": 4})
+/// ```
+///
+/// ## References
+/// - [Python documentation: Argument](https://docs.python.org/3/glossary.html#term-argument)
+#[violation]
+pub struct RepeatedKeywords {
+    duplicate_keyword: String,
+}
+
+impl Violation for RepeatedKeywords {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let dupe = &self.duplicate_keyword;
+        format!("Repeated keyword argument: `{dupe}`")
+    }
+}
+
+type KeywordRecordFn<'a> = Box<dyn FnMut(&str, TextRange) + 'a>;
+
+fn generate_record_func<'a>(checker: &'a mut Checker) -> KeywordRecordFn<'a> {
+    // init some hash sets to be captured by the closure
+    let mut seen = HashSet::<String>::new();
+    let mut dupes = HashSet::<String>::new();
+
+    let inner = move |keyword: &str, range| {
+        // Add an error the first time we see the duplicate
+        if seen.contains(keyword) && !dupes.contains(keyword) {
+            dupes.insert(String::from(keyword));
+            checker.diagnostics.push(Diagnostic::new(
+                RepeatedKeywords {
+                    duplicate_keyword: keyword.into(),
+                },
+                range,
+            ));
+        } else {
+            seen.insert(String::from(keyword));
+        }
+    };
+
+    Box::new(inner)
+}
+
+pub(crate) fn repeated_keywords(checker: &mut Checker, keywords: &Vec<Keyword>) {
+    let mut record_keyword = generate_record_func(checker);
+
+    for keyword in keywords {
+        if let Some(id) = &keyword.arg {
+            record_keyword(id.as_str(), keyword.range());
+        } else if let Expr::Dict(ExprDict {
+            // We only want to check dict keys if there is NO arg associated with them
+            keys,
+            range: _,
+            values: _,
+        }) = &keyword.value
+        {
+            for key in keys.iter().flatten() {
+                if let Expr::StringLiteral(ExprStringLiteral {
+                    value,
+                    range: _,
+                    unicode: _,
+                    implicit_concatenated: _,
+                }) = key
+                {
+                    record_keyword(value, key.range());
+                }
+            }
+        }
+    }
+}
```

Comment:
My advice would be to drop `generate_record_func`, and rewrite this as:

```rust
pub(crate) fn repeated_keywords(checker: &mut Checker, keywords: &Vec<Keyword>) {
    let mut seen =
        FxHashSet::with_capacity_and_hasher(keywords.len(), BuildHasherDefault::default());

    for keyword in keywords {
        if let Some(id) = &keyword.arg {
            if !seen.insert(id.as_str()) {
                checker.diagnostics.push(Diagnostic::new(
                    RepeatedKeywords {
                        duplicate_keyword: id.to_string(),
                    },
                    keyword.range(),
                ));
            }
        } else if let Expr::Dict(ExprDict {
            // We only want to check dict keys if there is NO arg associated with them
            keys,
            range: _,
            values: _,
        }) = &keyword.value
        {
            for key in keys.iter().flatten() {
                if let Expr::StringLiteral(ExprStringLiteral {
                    value,
                    range: _,
                    unicode: _,
                    implicit_concatenated: _,
                }) = key
                {
                    if !seen.insert(value.as_str()) {
                        checker.diagnostics.push(Diagnostic::new(
                            RepeatedKeywords {
                                duplicate_keyword: value.to_string(),
                            },
                            key.range(),
                        ));
                    }
                }
            }
        }
    }
}
```

There are a couple distinct changes here, so I'll walk through them one by one:

1. We tend to use `FxHashSet` over the standard library's `HashSet`. It tends to perform much better for smaller inputs.
2. You can initialize the hash set with a known capacity, which avoids resizing it unnecessarily for cases in which we know its approximate size.
3. In Rust, `seen.insert` returns a boolean (`false` if the element was already in the set). So we can both insert the element and test if it was already present in a single operation. The behavior here _is_ different than in your version, since we'll now flag the same argument multiple times if it's repeated more than twice, but honestly I think that's fine.
4. We can use a `FxHashSet<&str>` rather than an `FxHashSet<String>`, since we know the data will live longer than the hash set -- so there's no need to allocate a new String and have the hash set own the data. It can just use references.
5. I think the use of `Box`, `dyn`, etc. is not totally necessary since we're only calling this function twice, and after the refactor its been reduced to just a couple lines.

Happy to answer any questions!

Reply:
That makes a ton of sense. Thanks for such a detailed review! Needing to use the `Box` and `dyn` in what was logically a pretty simple function felt wrong, so I'm glad there was a more straightforward way of doing things